service/mapper/transitions.go

```diff
@@ -0,0 +1,380 @@
+
+	// At this point, we have collected all the payloads, so we go to the next
+	// step, where we will index them.
+	s.status = StatusCollected
+
+	return nil
+}
+
+func (t *Transitions) IndexRegisters(s *State) error {
+	log := t.log.With().Uint64("height", s.height).Hex("commit", s.next[:]).Logger()
+
+	// We should only index the payloads if we have just collected the payloads
+	// of a finalized block.
+	if s.status != StatusCollected {
+		return fmt.Errorf("invalid status for indexing registers (%s)", s.status)
+	}
+
+	// If there are no registers left to be indexed, we can go to the next step,
+	// which is about forwarding the height to the next finalized block.
+	if len(s.registers) == 0 {
+		log.Info().Msg("indexed registers for finalized block")
+		s.status = StatusIndexed
+		return nil
+	}
+
+	// We will now collect and index 1000 registers at a time. This gives the
+	// FSM the chance to exit the loop between every 1000 payloads we index. It
+	// doesn't really matter for badger if they are in random order, so this
+	// way of iterating should be fine.
+	n := 1000
+	paths := make([]ledger.Path, 0, n)
+	payloads := make([]*ledger.Payload, 0, n)
+	for path, payload := range s.registers {
+		paths = append(paths, path)
+		payloads = append(payloads, payload)
+		delete(s.registers, path)
+		if len(paths) >= n {
+			break
+		}
+	}
+
+	// Then we store the (maximum) 1000 paths and payloads.
+	err := t.index.Payloads(s.height, paths, payloads)
+	if err != nil {
+		return fmt.Errorf("could not index registers: %w", err)
+	}
+
+	log.Debug().Int("indexed", len(paths)).Int("remaining", len(s.registers)).Msg("indexed register batch for finalized block")
+
+	return nil
+}
+
+func (t *Transitions) ForwardHeight(s *State) error {
+
+	// We should only forward the height after we have just indexed the payloads
+	// of a finalized block.
+	if s.status != StatusIndexed {
+		return fmt.Errorf("invalid status for forwarding height (%s)", s.status)
+	}
+
+	// After finishing the indexing of the payloads for a finalized block, or
+	// skipping it, we should document the last indexed height. On the first
+	// pass, we will also index the first indexed height here.
+	var err error
+	t.once.Do(func() { err = t.index.First(s.height) })
+	if err != nil {
+		return fmt.Errorf("could not index first height: %w", err)
+	}
+	err = t.index.Last(s.height)
+	if err != nil {
+		return fmt.Errorf("could not index last height: %w", err)
+	}
+
+	// Now that we have indexed the heights, we can forward to the next height,
+	// and reset the forest to free up memory.
+	s.height++
+	s.forest.Reset(s.next)
+
+	t.log.Info().Uint64("height", s.height).Msg("height forwarded to next finalized block")
+
+	// Once the height is forwarded, we can set the status so that we index
+	// the blockchain data next.
+	s.status = StatusForwarded
+
+	return nil
+}
+
+func (t *Transitions) IndexChain(s *State) error {
+	log := t.log.Info().Uint64("height", s.height)
```

Comment:
Is it intentional that here we do:
```go
log := t.log.Info()
log.Msg("message")
```

while elsewhere we have:
```go
log := t.log.With().Str("field", value).Logger()
log.Info().Msg("message")
```

(I know the latter allows us to use different log levels)

Reply:
Yes.

In the first case, we conditionally keep adding fields to a single message.

In the second case, we have a shared set of fields that are always set, that will be part of any other message, while still being able to add additional fields.